File: packages/android/editor-schema/src/test/kotlin/app/notes/editor/ValidateTest.kt
package app.notes.editor

import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Test

class ValidateTest {
    @Test
    fun `accepts a minimal valid doc`() {
        validateDoc(Doc(listOf(DocNode(type = "paragraph"))))
    }

    @Test
    fun `rejects unknown block nodes`() {
        assertThrows(SchemaValidationException::class.java) {
            validateDoc(Doc(listOf(DocNode(type = "blockquote"))))
        }
    }

    @Test
    fun `rejects headings without a valid level`() {
        assertThrows(SchemaValidationException::class.java) {
            validateDoc(Doc(listOf(DocNode(type = "heading", attrs = mapOf("level" to 4)))))
        }
    }

    @Test
    fun `rejects lists with mismatched item types`() {
        assertThrows(SchemaValidationException::class.java) {
            validateDoc(
                Doc(
                    listOf(
                        DocNode(
                            type = "bullet_list",
                            content = listOf(
                                DocNode(
                                    type = "check_list_item",
                                    attrs = mapOf("checked" to false),
                                    content = listOf(DocNode(type = "paragraph")),
                                ),
                            ),
                        ),
                    ),
                ),
            )
        }
    }

    @Test
    fun `rejects check_list_item without boolean checked attr`() {
        assertThrows(SchemaValidationException::class.java) {
            validateDoc(
                Doc(
                    listOf(
                        DocNode(
                            type = "check_list",
                            content = listOf(
                                DocNode(
                                    type = "check_list_item",
                                    content = listOf(DocNode(type = "paragraph")),
                                ),
                            ),
                        ),
                    ),
                ),
            )
        }
    }

    @Test
    fun `rejects marks inside code_block`() {
        assertThrows(SchemaValidationException::class.java) {
            validateDoc(
                Doc(
                    listOf(
                        DocNode(
                            type = "code_block",
                            attrs = mapOf("language" to "js"),
                            content = listOf(
                                DocNode(type = "text", text = "hi", marks = listOf(Mark(MarkName.BOLD))),
                            ),
                        ),
                    ),
                ),
            )
        }
    }

    @Test
    fun `rejects empty text nodes`() {
        assertThrows(SchemaValidationException::class.java) {
            validateDoc(
                Doc(
                    listOf(
                        DocNode(
                            type = "paragraph",
                            content = listOf(DocNode(type = "text", text = "")),
                        ),
                    ),
                ),
            )
        }
    }

    @Test
    fun `rejects images without src`() {
        assertThrows(SchemaValidationException::class.java) {
            validateDoc(Doc(listOf(DocNode(type = "image", attrs = mapOf("alt" to "x")))))
        }
    }

    @Test
    fun `rejects attachments missing fields`() {
        assertThrows(SchemaValidationException::class.java) {
            validateDoc(Doc(listOf(DocNode(type = "attachment", attrs = mapOf("attachmentId" to "a")))))
        }
    }
}

File: packages/editor-schema/src/__tests__/validate.test.ts
import { describe, expect, it } from "vitest";
import { validateDoc } from "../validate.js";

describe("validateDoc", () => {
  it("accepts a minimal valid doc", () => {
    expect(() => validateDoc({ type: "doc", content: [{ type: "paragraph" }] })).not.toThrow();
  });

  it("rejects non-doc top-level types", () => {
    expect(() => validateDoc({ type: "paragraph", content: [] })).toThrow(/doc\.type/);
  });

  it("rejects unknown block nodes", () => {
    expect(() => validateDoc({ type: "doc", content: [{ type: "blockquote" }] })).toThrow(
      /unknown block/,
    );
  });

  it("rejects headings without a valid level", () => {
    expect(() =>
      validateDoc({
        type: "doc",
        content: [{ type: "heading", attrs: { level: 4 }, content: [] }],
      }),
    ).toThrow(/heading\.attrs\.level/);
  });

  it("rejects lists with mismatched item types", () => {
    expect(() =>
      validateDoc({
        type: "doc",
        content: [
          {
            type: "bullet_list",
            content: [
              {
                type: "check_list_item",
                attrs: { checked: false },
                content: [{ type: "paragraph" }],
              },
            ],
          },
        ],
      }),
    ).toThrow(/expected list_item/);
  });

  it("rejects check_list items without a boolean checked attr", () => {
    expect(() =>
      validateDoc({
        type: "doc",
        content: [
          {
            type: "check_list",
            content: [{ type: "check_list_item", content: [{ type: "paragraph" }] }],
          },
        ],
      }),
    ).toThrow(/checked must be a boolean/);
  });

  it("rejects code_block content with non-text nodes", () => {
    expect(() =>
      validateDoc({
        type: "doc",
        content: [
          { type: "code_block", attrs: { language: "js" }, content: [{ type: "paragraph" }] },
        ],
      }),
    ).toThrow(/text nodes/);
  });

  it("rejects marks on text inside code_block", () => {
    expect(() =>
      validateDoc({
        type: "doc",
        content: [
          {
            type: "code_block",
            attrs: { language: "js" },
            content: [{ type: "text", text: "hi", marks: [{ type: "bold" }] }],
          },
        ],
      }),
    ).toThrow(/marks are not allowed/);
  });

  it("rejects empty text nodes", () => {
    expect(() =>
      validateDoc({
        type: "doc",
        content: [{ type: "paragraph", content: [{ type: "text", text: "" }] }],
      }),
    ).toThrow(/text nodes need/);
  });

  it("rejects unknown marks", () => {
    expect(() =>
      validateDoc({
        type: "doc",
        content: [
          {
            type: "paragraph",
            content: [{ type: "text", text: "hi", marks: [{ type: "underline" }] }],
          },
        ],
      }),
    ).toThrow(/unknown mark/);
  });

  it("rejects duplicate marks on the same text node", () => {
    expect(() =>
      validateDoc({
        type: "doc",
        content: [
          {
            type: "paragraph",
            content: [{ type: "text", text: "hi", marks: [{ type: "bold" }, { type: "bold" }] }],
          },
        ],
      }),
    ).toThrow(/duplicate mark/);
  });

  it("rejects image nodes without a src", () => {
    expect(() =>
      validateDoc({
        type: "doc",
        content: [{ type: "image", attrs: { alt: "x" } }],
      }),
    ).toThrow(/image\.attrs\.src/);
  });

  it("rejects attachment nodes missing attachmentId or displayName", () => {
    expect(() =>
      validateDoc({
        type: "doc",
        content: [{ type: "attachment", attrs: { attachmentId: "a" } }],
      }),
    ).toThrow(/displayName/);
  });
});
